src/modules/navigator/mission.cpp

```diff
@@ -1033,19 +1047,54 @@ Mission::set_mission_items()
 				break;
 			}
 		}
+
+		if (_mission_item.nav_cmd == NAV_CMD_CONDITION_GATE) {
+			_mission_item.autocontinue = true;
+			_mission_item.time_inside = 0;
+		}
 	}
 
 	/*********************************** set setpoints and check next *********************************************/
 
 	position_setpoint_triplet_s *pos_sp_triplet = _navigator->get_position_setpoint_triplet();
 
-	/* set current position setpoint from mission item (is protected against non-position items) */
-	if (new_work_item_type != WORK_ITEM_TYPE_PRECISION_LAND) {
-		mission_apply_limitation(_mission_item);
-		mission_item_to_position_setpoint(_mission_item, &pos_sp_triplet->current);
+	// The logic in this section establishes the tracking between the current waypoint
+	// which we are approaching and the next waypoint, which will tell us in which direction
+	// we will change our trajectory right after reaching it.
+
+	// Because actions, gates and jump labels can be interleaved with waypoints,
+	// we are searching around the current mission item in the list to find the closest
+	// gate and the closest waypoint. We then store them separately.
+
+	// Check if the mission item is a gate
+	// along the current trajectory
+	if (item_contains_gate(_mission_item)) {
+
+		// The mission item is a gate, let's check if the next item in the list provides
+		// a position to go towards.
+
+		// TODO Precision land needs to be refactored: https://github.com/PX4/Firmware/issues/14320
+		if (has_next_position_item) {
+			// We have a position, convert it to the setpoint and update setpoint triplet
+			mission_apply_limitation(mission_item_next_position);
+			mission_item_to_position_setpoint(mission_item_next_position, &pos_sp_triplet->current);
+		}
+
+		// ELSE: The current position setpoint stays unchanged.
+
+	} else {
+		// The mission item is not a gate, set the current position setpoint from mission item (is protected against non-position items)
+		// TODO Precision land needs to be refactored: https://github.com/PX4/Firmware/issues/14320
+		if (new_work_item_type != WORK_ITEM_TYPE_PRECISION_LAND) {
+			mission_apply_limitation(_mission_item);
+			mission_item_to_position_setpoint(_mission_item, &pos_sp_triplet->current);
+		}
+
+		// ELSE: The current position setpoint stays unchanged.
 	}
 
-	/* only set the previous position item if the current one really changed */
+	// Only set the previous position item if the current one really changed
+	// TODO Precision land needs to be refactored: https://github.com/PX4/Firmware/issues/14320
```

Comment:
You've said that 3 times :smile:.

Reply:
Yes, in all the places where the fix needs to be applied.